### ISSUE TITLE
[MSI]: CLI and SharedFx bundle must set up the right registry keys for install detection

### ISSUE THREAD

**Issue by Sridhar-MS** [dotnet/cli#2588]
For example VC++ redist bundle sets up the correct registry keys for detection - https://github.com/dotnet/cli/blob/rel/1.0.0/packaging/windows/clisdk/bundle.wxs#L34

The CLI bundle and SharedFx bundle must do something similar. 

Note: CLI and SharedFx are installed SxS. So these registry keys must also be SxS and not overwrite each other.

cc @Petermarcu @joeloff

**Comment by Sridhar-MS** [comment]
@Petermarcu @robmen @joeloff @piotrpMSFT I like to get your feedback on this

Below is the registry I like to setup for CLI bundle detection

![image](https://cloud.githubusercontent.com/assets/9423246/14830226/7085255c-0ba4-11e6-8a25-98352ac9586c.png)

Since the CLI is installed SxS I am planning to set the registry keys in the following way
`HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\dotnet\Setup\$(platform)\CLI\$(cli_nugetversion) Value Install = 1`
Where
`$platform = x64 or x86`
`$cli_nugetversion = nuget version of the CLI (example - 1.0.0-rc2-002501)`

Some points to consider
- Both x64 and x86 installer will set the reg keys in the wow subsystem. This is to enable 32 bit installer to detect x64 or x86 installation.
- Since the CLI nuget version is part of the key any SxS installation will create a new key as shown in the above pic.
- CLI bundle detection logic for a give version example - `1.0.0-rc2-002501`
    - Check for the below reg key 
```xml
    <util:RegistrySearch
       Id="CLI_KeyExists"
       Variable="CLI_KeyExists"
       Root="HKLM"
       Key="SOFTWARE\dotnet\Setup\$(platform)\CLI\1.0.0-rc2-002501"
       Win64="no"
       Result="exists" />
```    
   
One disadvantage I see is that it becomes difficult to do a `greater than` detection. Similar to [here](https://github.com/dotnet/cli/blob/rel/1.0.0/packaging/windows/clisdk/bundle.wxs#L34). I am wondering that since CLI is SxS it does not make sense to do a `greater than` detection.

**Comment by joeloff** [comment]
Have you considered using the OptionalUpdateRegistration in the bundle? See http://wixtoolset.org/documentation/manual/v3/xsd/wix/optionalupdateregistration.html

**Comment by Petermarcu** [comment]
One of the things that's always been a challenge has been detecting that at least a certain version is installed. Do we want people parsing the key names and using semver to figure this out?

**Comment by Sridhar-MS** [comment]
> Have you considered using the OptionalUpdateRegistration

Seems like it is used for product updates

> This registration is intended primarily for update to an existing product.

But the CLI itself does not do any update during installation. It is strictly SxS. Are there any advantages  in using this? For example - does the bundle detection become much easier? Otherwise I see the registry keys are arranged in a similar way to my original proposal.

**Comment by joeloff** [comment]
We've been using it in our bundles simply as a means of tracking the bundle vs all the various MSIs we potentially install.

For the min version detection, isn't dotnet.exe shared across the CLIs? So if you do a filesearch and use that version you'll be able to find the highest version always, and if you want a specific version when there are multiple installs, you can rely on looking for specific registry keys.

**Comment by Petermarcu** [comment]
@joeloff , dotnet.exe won't be updated with each sdk version. They version independently.

**Comment by Sridhar-MS** [comment]
Maybe a registry value of type `REG_MULTI_SZ` can be used to store all the installed versions.. I am giving it a try so see if its feasible (especially if uninstalling will remove a version from that registry value).

**Comment by joeloff** [comment]
That could work, but it will still be difficult to parse the values. You'd likely need a custom action in an MSI and I'm not sure what you'd do at the bundle level to parse it.

**Comment by robmen** [comment]
Some thoughts off the top of my head:

1. Using semver pretty much guarantees a custom action. Semver just isn't deeply baked into the world of Windows yet (try to version a file with semver for example). As a starting point, it'd be good if we had a custom action for it and we've discussed adding support for semver to Burn in WiX v4.

2. Similarly, SxS everything isn't well supported today either. That's why you have "One of the things that's always been a challenge has been detecting that at least a certain version is installed." Again, I think we're looking at a custom action to teach the systems to understand how to enum through a registry key of versions or (if you must) MULTI_SZ. I'd recommend against MULTI_SZ since much of the world will just see that as a single key and then you're not SxS. Saying your SxS but having something that is not SxS just means you're in denial. :stuck_out_tongue_winking_eye: 

If you want to work with us in WiX to get the custom actions and "standard custom actions", we'd be thrilled to work with you. It'd be good to have semver supported (correctly) and a SxS pattern that worked.

**Comment by robmen** [comment]
PS: I still have no idea why I'm not getting any notifications when /cc:d on issues. Very frustrating.

**Comment by Sridhar-MS** [comment]
Yup. MULTI_SZ  is a bad idea..

**Comment by piotrpMSFT** [comment]
There is an open design issue here. @Sridhar-MS will set up a meeting to discuss...

**Comment by krwq** [comment]
# Requirements for CLI SDK and shared framework installer
- Support side by side installation and can be compatible with older versions of themselves
- Supports removing of any side by side version

## Scenarios
- Installing the product produces registry value informing about full version of the product which is being installed
- Installing the product produces registry values informing about compatible versions of the products
- Removing the product removes the registry keys if none of the versions of the product is supporting the given version. Specifically at least one registry key with full version should be removed.
- Installing older version of the product is possible

## Registry keys and values
- Full version is NuGet semantic version which contains the build number
- Version is NuGet semantic version which does *not* contain build number

Registry key should be created under following path:
```
HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\dotnet\Setup\<platform>\<installer-name>\[REG_DWORD]<nuget-version>=1
```

`<platform>=(x86|x64)`
`<installer-name>=(sdk|sharedfx)`

Example (installing CLI SDK x86 with version 1.0.3-rc2-123456):

```
HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\dotnet\Setup\x86\sdk\[REG_DWORD]1.0.0-rc2=1
HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\dotnet\Setup\x86\sdk\[REG_DWORD]1.0.1-rc2=1
HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\dotnet\Setup\x86\sdk\[REG_DWORD]1.0.2-rc2=1
HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\dotnet\Setup\x86\sdk\[REG_DWORD]1.0.3-rc2=1
HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\dotnet\Setup\x86\sdk\[REG_DWORD]1.0.3-rc2-123456=1
```

Note: 1.0.3 has two entries

Explanation:
CLI SDK version 1.0.3-rc2-123456 is compatible with any build of 1.0.3-rc2, 1.0.2-rc2, 1.0.1-rc2 and 1.0.0-rc2. If another installer is trying to find if there exists CLI SDK which supports specific version of installer they can check for presence of the value 1 under the registry path.

# Requirements for shared host installer
- Supports only in-place updated and is assumed to be always compatible with older versions of itself

## Scenarios
- Installing the product produces one registry value containing the latest version
- Installing older version on top of the other is not possible
- Removing removes the product completely with its registry keys, regardless of any previous installations

Registry key should be created under following path:
```
HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\dotnet\Setup\<platform>\[REG_SZ]sharedhost=<nuget-semversion>
```

# Example
```
HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\dotnet\Setup\x64\[REG_SZ]sharedhost=1.0.0-rc2-123456
```

**Comment by piotrpMSFT** [comment]
in case we need other values for an installer, can we create an `installedVersions` key between the `<installername>` and the version keys?

**Comment by piotrpMSFT** [comment]
When we're working on rc2 of 1.0.3, 1.0.1 and 1.0.2 would have already shipped. I'd expect that those versions would not include a suffix at this point in time.

**Comment by piotrpMSFT** [comment]
We don't want apps targeting RTM products to start running against a preview of the next RC. Imagine 1.0.1 has shipped. We don't want production apps targeting 1.0.1 to skip installation if 1.0.2-rc2 is present. I think that the "production back-compat" keys should be written until 1.0.2 reaches rtm itself.

**Comment by piotrpMSFT** [comment]
Can you show examples of a few scenarios?
- We install 1.0.0-rc2 to a clean machine
- We install first preview 1.0.0-rc3 installer and it installs on a clean machine
- We install preview 1.0.0-rc3 to a machine with RTM 1.0.0-rc2 installed
- We install 1.0.0 RTM to a clean machine
- We install 1.0.0 RTM to a machine with RC2 installed
- We install 1.0.0 RC2 preview to a machine with 1.0.0 RTM installed

**Comment by piotrpMSFT** [comment]
Fixed by #2828
